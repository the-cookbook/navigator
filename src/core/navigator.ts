import type { Method, Route, Parameters, QueryStrings, Navigator, MountNavigation } from '../contracts';
import toArray from '../utils/to-array';
import settings from '../settings';

import mountRoutePath from './mount-route-path';

const mountNavigation: MountNavigation = (path: string, queryStrings: QueryStrings, replace: boolean) => (
  params: Parameters = {},
  method: Method = 'push',
): void => {
  const mountedPath = mountRoutePath(path, params, queryStrings);

  if (!mountedPath) {
    return null;
  }

  if (replace || method === 'replace') {
    return settings.history.replace(mountedPath);
  }

  settings.history.push(mountedPath);
};

const navigator = (routes: Route[]): Navigator =>
  routes.reduce((acc: Navigator, { name, path, queryStrings = [], replace = false }: Route) => {
    acc[name] = mountNavigation(toArray(path)[0] as string, queryStrings, replace);

    return acc;
  }, {});

export default navigator;
