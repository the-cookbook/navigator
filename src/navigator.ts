import type { Route, Navigator } from './contracts';
import navigator from './core/navigator';
import settings from './settings';

const getAvailableRoutes = (): string =>
  settings.routes.reduce((list, route) => [...list, route.name], [] as Route[]).join(', ');

const shouldUpdate = (actual: Navigator): boolean =>
  !Object.keys(actual).length || Object.keys(actual).length !== Object.keys(settings.routes).length;

export default new Proxy(
  {},
  {
    get(target: Navigator, route: string) {
      if (shouldUpdate(target)) {
        // eslint-disable-next-line no-param-reassign
        target = navigator(settings.routes);
      }

      if (!target[route]) {
        throw new RangeError(
          `\n\nRoute "${route}" is not available in declared routes:\n\nâ€¢ "${getAvailableRoutes()}"\n`,
        );
      }

      return target[route];
    },
  },
);
