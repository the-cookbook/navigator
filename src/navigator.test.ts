import navigator from './navigator';

jest.mock('./settings');

describe('navigator', () => {
  it('should throw error when route is not available', () => {
    const to = 'non-existent-route';

    expect(() => navigator[to]).toThrowErrorMatchingSnapshot();
  });

  it('should not throw error when route exists', () => {
    const to = 'profile';

    expect(navigator[to]).not.toThrowError();
  });
});
