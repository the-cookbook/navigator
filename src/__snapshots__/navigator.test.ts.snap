// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigator should throw error when route is not available 1`] = `
"

Route \\"non-existent-route\\" is not available in declared routes:

â€¢ \\"home, profile, search, subscriptions, internalServerError, notFound\\"
"
`;
